Fix pipeline stall in `do` during capture and remove excessive redirections

SUPERCILEX opened the pull request:
Currently, if you run `do -i { sudo apt upgrade }`, stdin gets swallowed and doesn't let you respond yes/no to the upgrade question. This PR fixes that, but runs into https://github.com/nushell/nushell/issues/7205 so the tests fail.

sholderbach commented:
Well neither a pipeline stalll nor discarding stdin is great. (Is #7205 reproducible outside of this branch?)

SUPERCILEX commented:
~Yes. A little bit of explanation: the stdin swallowing is caused by redirecting stderr in `eval_block`. When I fixed that, it broke the tests that were using `-i` for capturing (which I'm pretty sure is wrong). So then I tried to change them to use `-c`, but there's a pipeline stall. I also double checked and this behavior is reproducible before #7122.~

Figured it out, will post and updated patch + explanation later today.

SUPERCILEX commented:
Ok, I thought I had it, but not quite. I think you guys need to fundamentally rework streams to get this right. Anyway, this PR at least fixes the swallowing and pipeline freezes caused by #7122.

Details: it turns out I did break some stuff in #7122, so I had to bring back the stderr/stdout copying. Based on poking around some older commits, it seems like the order in which the streams is consumed is important. There's still one test I couldn't figure out (`capture_error_with_too_much_stdout_not_hang_nushell`), but it passes as it was written before this PR (with `-i`) so it's probably a more fundamental issue.

WindSoilder commented on crates/nu-command/tests/commands/do_.rs:
```diff
@@ -122,8 +122,9 @@ fn ignore_error_with_too_much_stderr_not_hang_nushell() {
 }
 
 #[test]
+#[ignore]
```
You don't need to ignore the test, I think it can be fixed easily

SUPERCILEX replied:
Done.

WindSoilder commented on crates/nu-command/src/core_commands/do_.rs:
```diff
@@ -117,6 +118,26 @@ impl Command for Do {
+                };
+
+                let mut stdout_ctrlc = None;
+                let stdout_msg = match stdout {
```
you can delegate the stdout stream consuming login in another thread, then we can avoid deadlock

WindSoilder replied:
You can check my pr(#7234) and do something similar:
```rust
                let stdout_handler = stdout.map(|stdout_stream| {
                    std::thread::spawn(move || {
                        let stdout_ctrlc = stdout_stream.ctrlc.clone();
                        let stdout_span = stdout_stream.span;
                        let stdout_bytes = match stdout_stream.into_bytes() {
                            Err(_) => vec![],
                            Ok(bytes) => bytes.item,
                        };
                        RawStream::new(
                            Box::new(vec![Ok(stdout_bytes)].into_iter()),
                            stdout_ctrlc,
                            stdout_span,
                        )
                    })
                });

               // then consuming stderr message

                ....
                ....

                let stdout = if let Some(h) = stdout_handler {
                    match h.join() {
                        Err(err) => {
                            return Err(ShellError::ExternalCommand(
                                "Fail to receive external commands stdout message".to_string(),
                                format!("{err:?}"),
                                span,
                            ))
                        }
                        Ok(res) => Some(res),
                    }
                } else {
                    None
                };
```

SUPERCILEX replied:
Thanks! I still think the stream framework needs to be rewritten so this bug is impossible, but this is a good stopgap solution.

WindSoilder commented on crates/nu-command/tests/commands/do_.rs:
```diff
@@ -138,7 +139,7 @@ fn ignore_error_with_too_much_stdout_not_hang_nushell() {
             cwd: dirs.test(), pipeline(
             r#"
-                do -i {sh -c "cat a_large_file.txt"} | complete | get stdout
+                do -c {sh -c "cat a_large_file.txt"} | complete | get stdout
```
I think it's better to add three more test cases about `do -c` rather than change existing test

SUPERCILEX replied:
I'd agree, except that `-i` no longer captures so these tests won't work with `-i`.

rgwood commented:
Thanks for the PR!

I'm not totally sure where this PR is at now. Sounds like it fixes multiple issues now.

Would you be able to provide some quick examples of Nu code that did not work before this PR that work now?

SUPERCILEX commented:
Yeah, no prob. It fixes two intertwined things:
- `do -i { sudo apt upgrade }` was broken (if you run it, "Do you want to continue?" will abort instead of letting you answer), but works now because we stop redirecting stderr when ignoring errors.
- Because we stopped capturing errors on `-i`, the tests that relied on that behavior had to be updated to actually capture errors. That led to the discovery of a deadlock that was fixed by consuming stderr and stdout on different threads.

rgwood commented:
Thanks! I am trying to review the code changes, and I'm having a difficult understanding _why_ the changes fix those problems. Any code comments or explanations you can offer would be much appreciated.

SUPERCILEX commented:
Neither do I unfortunately. 😅 @WindSoilder might know more? I just know that stuff doesn't work before this PR but does afterwards.

WindSoilder commented:
I think it's not cature stdin, it's actually capture stderr message, and you should be ok to type `yes/no` to response to `apt`

SUPERCILEX commented:
@WindSoilder for me, apt would immediately print "Abort" and wouldn't actually let me input anything (this is weird, I know).

Taking a step back: stderr weirdness aside, I think this PR is a good idea on its own. It doesn't make sense to capture stderr when ignoring errors (unless you specify the capture flag).

SUPERCILEX commented:
Something just occurred to me: I was testing apt with the non-threaded capturing, but I wonder if that was the issue and threaded capturing fixes it. Won't be able to check that for a few days as I need to be in Windows land.

WindSoilder commented:
> apt would immediately print "Abort" and wouldn't actually let me input anything (this is weird, I know)

So I guess that's the `apt`'s implementation detail:  `apt` itself is trying to find if it captures `tty`, if not, just abort itself.  (Here is my guess, correct me if I'm wrong)

WindSoilder commented:
> apt would immediately print "Abort" and wouldn't actually let me input anything (this is weird, I know)

So I guess that's the `apt`'s implementation detail:  `apt` itself is trying to find if it captures `tty`, if not, just abort itself.  (Here is my guess, correct me if I'm wrong)

rgwood commented:
> stderr weirdness aside, I think this PR is a good idea on its own

It seems like the PR is solving real problems, but we need to get to a point where we confidently understand _why_ it's solving those problems.

For changes like this where reading the code is not enough, we would like to see some comments explaining its purpose and operation.

SUPERCILEX commented:
> > apt would immediately print "Abort" and wouldn't actually let me input anything (this is weird, I know)
> 
> So I guess that's the `apt`'s implementation detail: `apt` itself is trying to find if it captures `tty`, if not, just abort itself. (Here is my guess, correct me if I'm wrong)

That would actually make a lot of sense.

> For changes like this where reading the code is not enough, we would like to see some comments explaining its purpose and operation.

That's fair: can you share what you're specifically confused about? The apt stuff is weird, but the rest makes sense IMO. Not really sure what I should be doing on my end.

rgwood commented:
> That's fair: can you share what you're specifically confused about?

I don't understand the purpose of any of the code you added to `do_.rs`.

WindSoilder commented:
Oh... After thinking about this, I've one more concern for it, I think the change will `complete` usage.  Take the following script as example:
```
# test.py
import sys

print('aaa', file=sys.stderr)
```

Run `do -i { python test.py } | complete` in the pr branch:
```
❯ do -i {python test.py} | complete
aaa
╭────────┬──╮
│ stdout │  │
╰────────┴──╯
```

Current main:
```
❯ do -i {python test.py} | complete
╭────────┬─────╮
│ stdout │     │
│ stderr │ aaa │
│        │     │
╰────────┴─────╯
```

I think the better way to handle this is change the `apt` usage, like `do -i {sudo apt upgrade -y }`

SUPERCILEX commented:
@rgwood thanks, that helps! I'll add some comments.

@WindSoilder that's intentional though. If you wanted to capture, you should use `-c`. Or are you saying that `-i` is also supposed to capture? That doesn't make sense to me.

WindSoilder commented:
> you should use -c

But `-c` is not capture either, if our command runs to failed, it will raise out error, and `complete` command has no chance to collect stderr message.

theAkito commented:
From what I have read on the documentation, experienced when playing around with Nu & thought about when drawing conclusions, this is what _should_ be the case.


```nu
> do { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" }            
good
bad

> $env.LAST_EXIT_CODE
1
```

Standard behaviour without `complete`.

```nu
do { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
┏━━━━━━━━━━━┳━━━━━━┓
┃ stdout    ┃ good ┃
┃           ┃      ┃
┃ stderr    ┃ bad  ┃
┃           ┃      ┃
┃ exit_code ┃ 1    ┃
┗━━━━━━━━━━━┻━━━━━━┛
```

Standard behaviour with `complete`.

```nu
do -i { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
┏━━━━━━━━━━━┳━━━━━━┓
┃ stdout    ┃ good ┃
┗━━━━━━━━━━━┻━━━━━━┛
```

Ignoring all errors.

```nu
do -c { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" }
good
Error: nu::shell::external_command (link)

  × External command failed
   ╭─[entry #24:1:1]
 1 │ do -c { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" }
   ·         ─┬
   ·          ╰── External command failed
   ╰────
  help: bad
```

Executing & raising exception, if something goes wrong.


```nu
do -c { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
Error: nu::shell::external_command (link)

  × External command failed
   ╭─[entry #3:1:1]
 1 │ do -c { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
   ·         ─┬
   ·          ╰── External command failed
   ╰────
  help: bad
```

Executing & raising exception, if something goes wrong, but `stdout` is swallowed by complete. Cannot be accessed due to exception, but that's another story.

Exceptions are useful for `try` `catch`.

With this definition, the `-c` ought to be changed to something that does not imply "capturing" errors.

The definition is in my view the most reasonable, intuitive, simple & best to work with. Please, tell me if I'm missing something. This `do` topic is pretty confusing.

The most important paragraph in this definition is that the standard `do` must "capture" everything and output everything. The default behaviour ought to not swallow anything.

SUPERCILEX commented:
Hmmm, that's a good point about redirect as a separate command—I like that quite a bit actually. So is the conclusion that `do` shouldn't do any redirection at all? That seems reasonable to me.

WindSoilder commented:
I'm not sure why using `redirect` command instead of current version of builtin redirection like `out> 1.txt`, or `err> 2.err.txt`.

SUPERCILEX commented:
Ok, had some time to look at this again. @WindSoilder I agree that we want the `complete` behavior that @theAkito described, but using redirection is a hack IMO. If you set `redirect_stderr` to true in `eval_block_with_early_return`, `... | complete` will work as expected and have a stderr row, but `apt upgrade` will be broken. I think what's going on is that redirect in the eval.rs context means "don't send me to the console" as in redirect into something nu has control over. That's the behavior that's broken: `complete` should be able to back propagate the request for stderr to not be sent to the console, but we shouldn't be hacking `do` so that it sometimes works with complete and sometimes doesn't. I think this might require something like the transformations API I mentioned because currently every command seems to decide on its own whether or not to send output to the console, but `complete` needs a way to take over the entire pipeline.

Anyway, for this PR I added the requested comments.

rgwood commented:
Thanks for adding the detailed comments.

I appreciate the effort on this PR but I'm struggling to review it; I don't have a great understanding of the current state of `do` nor where we want it to be.

Hopefully someone else has a better understanding of this area and is able to help make a decision.

WindSoilder commented on crates/nu-command/src/core_commands/do_.rs:
```diff
@@ -106,7 +109,8 @@ impl Command for Do {
             call.redirect_stdout,
-            capture_errors || ignore_shell_errors || ignore_program_errors,
+            // TODO remove capture_errors once redirection is fixed
+            call.redirect_stderr || capture_errors,
```
I think in nushell `call.redirect_stderr` will always be `false`.

I'd suggest change the line to

`capture_errors || !ignore_all_errors`, so `do -i` will not redirect stdout, and standard `do` without `-i` will capture stderr and have a chance to collect by `complete`.

WindSoilder reviewed (commented):
Thanks for your change, I just leave one thought, otherwise everything looks fine to me.

WindSoilder commented on crates/nu-command/src/core_commands/do_.rs:
```diff
@@ -106,7 +109,8 @@ impl Command for Do {
             call.redirect_stdout,
-            capture_errors || ignore_shell_errors || ignore_program_errors,
+            // TODO remove capture_errors once redirection is fixed
+            call.redirect_stderr || capture_errors,
```
I think `call.redirect_stderr` will always be false, so it's meanless

I'd suggest change the line to: `capture_errors || !ignore_all_errors`, so do -i will not redirect stdout, and standard do without -i will capture stderr and have a chance to collect by complete.

SUPERCILEX replied:
Doesn't that mean apt and friends will be broken if you stick them in a plain do block? I still fundamentally disagree with the idea of making this stuff sometimes work and sometimes not depending on how you massage it.

WindSoilder replied:
yeah, but it works for `do -i`, One of `do` and `do -i` must capture stderr.

I think we made a decision that `do -i` should made nushell ignore stderr message(so it can be printed), so plain do should capture stderr and can be catched by `complete` command

WindSoilder replied:
Refer to comment: https://github.com/nushell/nushell/pull/7204#issuecomment-1337896214

I still want `do` to do something like @theAkito mentioned.
```
> do { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" }            
good
bad

> $env.LAST_EXIT_CODE
1
```
Standard behaviour without complete(This can't be completed, but we can achieve the behavior with `do -i`).

```
do { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
┏━━━━━━━━━━━┳━━━━━━┓
┃ stdout    ┃ good ┃
┃           ┃      ┃
┃ stderr    ┃ bad  ┃
┃           ┃      ┃
┃ exit_code ┃ 1    ┃
┗━━━━━━━━━━━┻━━━━━━┛
```
Standard behaviour with complete(this will be ok).

```
do -i { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
┏━━━━━━━━━━━┳━━━━━━┓
┃ stdout    ┃ good ┃
┗━━━━━━━━━━━┻━━━━━━┛
```
Ignoring all errors(this will be ok).

SUPERCILEX replied:
Yeah, I totally agree that `do { blah } | complete` should work, but I'm saying it shouldn't work at the expense of breaking blah (which is currently the case AFAIK). Maybe this is something that should be moved to an issue about managing redirection with complete so it can be fixed properly?

WindSoilder replied:
Ok, I agree with you, I'd vote for landing it.

WindSoilder replied:
Randomly thinking, if `do { blah } | complete` works and complete catch `stderr`, `blah` must redirect `stderr`.

I think ideally what we can reach is make `do { blah }` works without redirecting `stderr`, and `do { blah } | complete` works, but  `blah` must redirecting `stderr`

WindSoilder commented:
Randomly thinking, if `do { blah } | complete` works and *complete* catch stderr, blah must redirect stderr, I don't think there is a way to get rid of it.

I think ideally what we can reach is:
1. `do { blah }` works without redirecting stderr, and,
2. `do { blah } | complete` works too, but blah must redirecting stderr

theAkito commented:
> Randomly thinking, if `do { blah } | complete` works and _complete_ catch stderr, blah must redirect stderr, I don't think there is a way to get rid of it.
> 
> I think ideally what we can reach is:
> 
>     1. `do { blah }` works without redirecting stderr, and,
> 
>     2. `do { blah } | complete` works too, but blah must redirecting stderr
> 
> 
> @SUPERCILEX What do you think about this?

Sounds sound. 😄

WindSoilder commented:
https://github.com/nushell/nushell/blob/80463d12fb861c227e1492bc83fd4df003c4fabc/crates/nu-command/src/core_commands/do_.rs#L109

Sorry, I'd like to make a final suggestion: change the line to `!ignore_all_errors && (call.redirect_stdout || capture_errors)`.

Then:
- `do -i { blah }`, `do { blah }`, `do -i { blah } | complete` work without `blah` stderr redirection.
- `do { blah } | complete` would make `blah` redirect stderr.

It matches all the cases in https://github.com/nushell/nushell/pull/7204#issuecomment-1337896214 (except `capture_errors` flag, but that's another story)

It's a relative simple change, for long term, it's better to check if the next command after `do` is actually `complete`, then change `call.redirect_stderr` to true.

SUPERCILEX commented:
Wait a minute!!! `call.redirect_stdout` is talking about the *next* command? I completely missed that. Hopefully I'm not forgetting something, but why are we not just setting stderr redirect to `call.redirect_stdout`. That's super simple:
- Plain `do` with no piping always works.
- Any piping always redirects stderr.

It's actually what `apt upgrade` is trying to detect anyway: if you pipe it, then a user isn't going to see the output it requires an explicit `--yes`, otherwise the user can see it so make the command interactive.

SUPERCILEX commented:
The one wrinkle I can't figure out:
- In bash,
```bash
$ bash -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" | cat
bad
good
```
You can see that `bad` is still printed.
- For some reason, the same thing in nu:
```nu
> do { bash -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | cat
good
```
doesn't print `bad` which I don't understand because it shows up in complete so clearly stderr is there:
```
〉do { bash -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete       01/09/2023 06:41:00 PM
╭───────────┬──────╮
│ stdout    │ good │
│           │      │
│ stderr    │ bad  │
│           │      │
│ exit_code │ 1    │
╰───────────┴──────╯
```
Of course if you remove the piping it works since stderr isn't being redirect anymore.

If we can figure that out it'd be great, but I'm actually happy with this PR now. WDYT?

WindSoilder commented:
Yeah I just found that `call.redirect_stdout` is useful for our case.

> Any piping always redirects stderr.

So `do -i` will redirect and complete will capture stderr too.
```
❯ do -i { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | complete
╭────────┬──────╮
│ stdout │ good │
│        │      │
│ stderr │ bad  │
│        │      │
╰────────┴──────╯
```

And plain `do -i` without complete will print both good and bad
```
❯ do -i { sh -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" }
good
bad
```

@theAkito What do you think about it?

WindSoilder commented:
```
> do { bash -c "function bad { echo good; echo bad 1>&2; return 1; }; bad" } | cat
good
```
doesn't print bad which I don't understand because it shows up in complete so clearly stderr is there

Yeah I think this can be improved in the future, we can only redirect stderr **only if** the next command is `complete`.

SUPERCILEX commented:
> Yeah I think this can be improved in the future, we can only redirect stderr only if the next command is complete.

Yeep, sounds good.

rgwood commented:
Let's merge this and try it out; @WindSoilder has done a detailed review and thinks it looks good.

Thanks for the fix!

fdncred commented:
I can't get any of these tests to work. They all hang in macos. We might want to consider reverting this PR.
```
capture_error_with_too_much_stderr_not_hang_nushell()
capture_error_with_too_much_stdout_not_hang_nushell()
capture_error_with_both_stdout_stderr_messages_not_hang_nushell()
```

SUPERCILEX commented:
Maybe run a clean? CI runs the tests on macOS, so it's super weird that they'd fail.

fdncred commented:
> Maybe run a clean? CI runs the tests on macOS, so it's super weird that they'd fail.

@WindSoilder helped me solve it. Everything wasn't building properly.

fdncred commented:
Darn, I'm troubleshooting something else now. My oh-my.nu script isn't working right now because of `do -i` changes. Not sure if it's related to this PR or what.

```
def get_icons_list [] {
    {
        AHEAD_ICON: (char branch_ahead), # "↑" 2191
        BEHIND_ICON: (char branch_behind), # "↓" 2193
        NO_CHANGE_ICON: (char branch_identical) # ≣ 2263
        HAS_CHANGE_ICON: "*",
        INDEX_CHANGE_ICON: "♦",
        WT_CHANGE_ICON: "✚",
        CONFLICTED_CHANGE_ICON: "✖",
        UNTRACKED_CHANGE_ICON: (char branch_untracked) # ≢ 2262
        INSERT_SYMBOL_ICON: "❯",
        HAMBURGER_ICON: (char hamburger) # "≡" 2261
        GITHUB_ICON: "", # f408
        BRANCH_ICON: (char nf_branch) # "" e0a0
        REBASE_ICON: "", # e728
        TAG_ICON: "" # f412
    }
}

def get_icon_by_name [name] {
    do -i { get_icons_list | get $name }
}
```

`get_icon_by_name NO_CHANGE_ICON` should return the hamburger and it doesn't any longer.

SUPERCILEX commented:
Why are you ignoring errors? That's probably hiding a crash.

fdncred commented:
If I'm doing this with a valid $name
```
do -i { get_icons_list | get $name }
```
shouldn't it work the same as
```
get_icons_list | get $name
```

SUPERCILEX commented:
Hmmm, yeah. Are you sure that's related to this PR? Unless redirection broke things, there shouldn't be any relevant changes.

WindSoilder commented:
Sorry, my bad, I found there is a line of change breaks it:
```rust
            Ok(PipelineData::Value(..)) | Err(_) if ignore_shell_errors => {
                Ok(PipelineData::empty())
            }
```

It eats all valid values like PipelineData::Value::Int(3)

SUPERCILEX commented:
Crap, that's embarrassing, my bad. Fixed here: https://github.com/nushell/nushell/pull/7859.

fdncred commented:
Is this another bug or something else?

Shouldn't we be able to eat the errors here?

I tried both of these.
```
help commands | find arg | get search_terms | do --ignore-errors { ansi strip }
help commands | find arg | get search_terms | do --ignore-program-errors { ansi strip }
```
And they both print this
```
Error: nu::shell::type_mismatch (link)

  × Type mismatch.
   ╭─[entry #75:1:1]
 1 │ help commands | find arg | get search_terms | do --ignore-shell-errors { ansi strip }
   · ──────┬──────
   ·       ╰── value is nothing, not string
   ╰────
```

SUPERCILEX commented:
I don't know which part of the pipeline is failing, but you probably need to wrap more of it in the do block?

fdncred commented:
I wrapped the whole thing and it gives the same error.

SUPERCILEX commented:
Huh, well `ansi` is correctly returning an error type: https://github.com/nushell/nushell/blob/731f5f852304358c83abb4595b6d80e9156b31a5/crates/nu-command/src/platform/ansi/strip.rs#L85-L89

But the `do_` command is seeing an Ok:
```
[crates/nu-command/src/core_commands/do_.rs:115] &result = Ok(
    ListStream(
        ListStream,
        None,
    ),
)
```

Oddly enough if you append a `| echo` to the end the command doesn't fail at all. Don't have time to debug further than that, but I don't think it's related to `do`.

WindSoilder commented:
Hi, @fdncred I wouldn't consider it causes the issue you provides above, because I can reproduce it in 0.74.0, the latest release
```
❯ help commands | find arg | get search_terms | do -i { ansi strip }
Error: nu::shell::type_mismatch (link)

  × Type mismatch.
   ╭─[entry #3:1:1]
 1 │ help commands | find arg | get search_terms | do -i { ansi strip }
   · ──────┬──────
   ·       ╰── value is nothing, not string
   ╰────

❯ version | transpose key value | to md --pretty
| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.74.0                                   |
| branch             |                                          |
| commit_hash        | 9a56665c6beee31e365ce17898056f1ff2db4094 |
| build_os           | macos-x86_64                             |
| build_target       | x86_64-apple-darwin                      |
| rust_version       | rustc 1.65.0 (897e37553 2022-11-02)      |
| rust_channel       | 1.65.0-x86_64-apple-darwin               |
| cargo_version      | cargo 1.65.0 (4bc8f24d3 2022-10-20)      |
| pkg_version        | 0.74.0                                   |
| build_time         | 2023-01-10 20:37:38 +00:00               |
| build_rust_channel | release                                  |
| features           | database, default, trash, which, zip     |
| installed_plugins  | from-binary                              |
```

fdncred commented:
Thanks @SUPERCILEX and @WindSoilder for investigating. I'm glad it's not this PR. I'll open a new issue.
